stacks-signer/src/v0/signer_state.rs

```diff
@@ -304,50 +299,122 @@ impl LocalStateMachine {
+        )?;
+        state_machine.update_time = UpdateTime::now();
+        info!(
+            "Signer State: Current tenure timed out, setting the active miner to the prior tenure";
+            "inactive_tenure_ch" => %inactive_tenure_ch,
+            "new_active_tenure_ch" => %new_active_tenure_ch
+        );
+
+        crate::monitoring::actions::increment_signer_agreement_state_change_reason(
+            crate::monitoring::SignerAgreementStateChangeReason::InactiveMiner,
+        );
+
+        Ok(())
+    }
+
+    /// Retrieves the last known block height and ID for a given parent tenure, querying
+    /// both the connected stacks-node and local signer database to determine the most
+    /// recent block associated with the specified `parent_tenure_id`.
+    pub fn get_parent_tenure_last_block(
+        client: &StacksClient,
+        db: &SignerDb,
+        tenure_last_block_proposal_timeout: Duration,
+        parent_tenure_id: &ConsensusHash,
+    ) -> Result<(u64, StacksBlockId), SignerChainstateError> {
+        let stacks_node_last_block = client
+            .get_tenure_tip(parent_tenure_id)
+            .inspect_err(|e| {
+                warn!(
+                    "Signer State: Failed to fetch last block in parent tenure from stacks-node";
+                    "parent_tenure_id" => %parent_tenure_id,
+                    "err" => ?e,
+                )
+            })
+            .ok()
+            .map(|header| {
+                (
+                    header.height(),
+                    StacksBlockId::new(parent_tenure_id, &header.block_hash()),
+                )
+            });
+        let signerdb_last_block = SortitionData::get_tenure_last_block_info(
+            parent_tenure_id,
+            db,
+            tenure_last_block_proposal_timeout,
+        )?
+        .map(|info| (info.block.header.chain_length, info.block.block_id()));
+
+        let (parent_tenure_last_block_height, parent_tenure_last_block) =
+            match (stacks_node_last_block, signerdb_last_block) {
+                (Some(stacks_node_info), Some(signerdb_info)) => {
+                    std::cmp::max_by_key(stacks_node_info, signerdb_info, |(chain_length, _)| {
+                        *chain_length
+                    })
+                }
+                (None, Some(signerdb_info)) => signerdb_info,
+                (Some(stacks_node_info), None) => stacks_node_info,
+                (None, None) => {
+                    return Err(SignerChainstateError::NoParentTenureInfo(*parent_tenure_id))
+                }
+            };
+        Ok((parent_tenure_last_block_height, parent_tenure_last_block))
     }
 
     fn make_miner_state(
-        sortition_to_set: SortitionState,
+        sortition_to_set: SortitionData,
         client: &StacksClient,
         db: &SignerDb,
-        proposal_config: &ProposalEvalConfig,
+        tenure_last_block_proposal_timeout: Duration,
     ) -> Result<MinerState, SignerChainstateError> {
```

Comment:
Should `make_miner_state` be deduped with the new `get_parent_tenure_last_block`?